### DISCUSSION TITLE
Not exiting code

### DISCUSSION THREAD

**Discussion by infinite-pursuits** [Consensys/gnark#910] [Q&A]
I have a circuit with an api.AssertIsEqual statement. When this constraint is not met, I get an error and the code is exited. Is there a way to continue running the code despite the circuit constraints not being satisfied? like a try catch block in python.

**Comment by ivokub** [comment]
gnark doesn't have such option. What are you trying to achieve?

When testing circuits then I suggest using `test.IsSolved` method. This performs all the operations inline on `*big.Int` types and is significantly faster than running the setup and prover every time.

**Reply by infinite-pursuits** [reply]
I just wish to run the rest of my code even if the circuit fails.

**Reply by ivokub** [reply]
It is not possible and wouldn't even make sense - if the circuit fails then the witness would be invalid. It is not possible to generate a proof for invalid witness.

Regarding test engine - adding such functionality would be difficult. We panic in case assertions do not hold to be compatible with `frontend.API`. It is possible to recover from the panic, but it would be really difficult to re-continue where we left off. It would mean that we would have to parse the circuit source code and start executing from partial circuit meanwhile having to restore the internal state. Development effort and maintenance overhead definitely wouldn't balance the feature for which there is a simple workaround (commenting out the part which fails in circuit definition).
